### PR TITLE
[FEATURE] Add DataPlotly algorithm to processing : build simple plot

### PR DESCRIPTION
**Pull request by mdouchin**
First processing alg added which allows to build a simple plot. Basic input parameters:

* Layer
* Plot type : bar, pie, scatter, histogram
* X field
* Y field
* Output HTML file
* Output JSON data and layout of the plot

### CODE REVIEW

File: processing/dataplotly_algorithms.py [hunk 1]
```diff
@@ -49,6 +49,7 @@ class DataPlotlyProcessingPlot(QgisAlgorithm):
 
     INPUT = 'INPUT'
     PLOT_TYPE = 'PLOT_TYPE'
+    PLOT_TITLE = 'PLOT_TITLE'
     PLOT_TYPE_OPTIONS = ['bar', 'pie', 'scatter', 'histogram']
```
Comment (ghtmtt) [inline processing/dataplotly_algorithms.py]: maybe it is worth to keep the order of the plugin plot (scatter, bar, histogram, pie)

File: data_plotly_plot.py [hunk 2]
```diff
@@ -58,6 +58,80 @@ class Plot(object):
 
     def __init__(self, plot_type, plot_properties, plot_layout):
 
+        # Define default plot dictionnary used as a basis for plot initilization
```
Comment (ghtmtt) [inline data_plotly_plot.py]: good idea to define in advance the dictionary with all the plot and layout properties. This means that also in the future, when adding new plot or layout parameters we have to add them also here, isn't it?

**Comment by ghtmtt** [comment]
@mdouchin many many thanks for this PR. I think all the changes are OK: I've made some random tests to be sure that the basic functionalities of the plugin work. 

Creating the basic plot and layout parameters in the Plot class is smart: though this changes a little bit the adding of other parameters in the future (I've to remember to add also the empty parameter there ;) ). 

As I comment above, I cannot see the html and json output in the result viewer when it is saved as temporary output and I think that, if possible, we have to add them also there. 

One more minor stuff is for the scatterplot: if not specified, plotly creates the scatter plot with both line and marker but I think most of the users will have by default only the markers (that is `mode = 'markers'` in the plot parameters). Could you maybe change it and *force* the scatter plot to behave like that? 

It seems also that it is mandatory to specify a plot title: when nothing is typed Processing throws an error and no plot is generated. Maybe we have to let this as an optional parameter?

When histogram is chosen both X and Y are available but histogram wants only the X parameter. I think having both X and Y could confuse the user especially because in the legend of the resulting histogram there isn't specified the name of the variable.


This is my first review of a PR and I'm so glad I could made it with you really! :) As you can see there are IMHO just some minor and small stuff to change. 

Thanks again for taking care of this feature!

File: processing/dataplotly_algorithms.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    Create a simple plot with DataPlotly plugin
+    """
+
+    # Constants used to refer to parameters and outputs. They will be
+    # used when calling the algorithm from another algorithm, or when
+    # calling from the QGIS console.
+
+    INPUT = 'INPUT'
+    PLOT_TYPE = 'PLOT_TYPE'
+    PLOT_TITLE = 'PLOT_TITLE'
```
Comment (ghtmtt) [inline processing/dataplotly_algorithms.py]: maybe we have to let this parameter optional

**Comment by mdouchin** [comment]
Thanks @ghtmtt I will read your review and improve my PR asap

**Comment by mdouchin** [comment]
Commit 7d5a2fe set the title optionnal

**Comment by mdouchin** [comment]
> Creating the basic plot and layout parameters in the Plot class is smart: though this changes a little bit the adding of other parameters in the future (I've to remember to add also the empty parameter there ;)

I think we should really find a way to unite the information about the plot types. For example, we could use the same dict to store where each widget belongs too, and which plotly parameters are concerned by which plot type. `self.widgetType`. Something like

```python
plotBaseProperties = {
    'x': { 'default' : None, 'plot_types': ['all'] }
    'y': { 'default' : None, 'plot_types': ['scatter', 'bar', 'box', 'pie', '2dhistogram', 'polar', 'ternary', 'contour'] },
     'z': { 'default' : None, 'plot_types': ['ternary'] }
}
```

Doing so, we could use this dict in bot purpose. Note that you can get a dialog object (ex: QComboBox) by its name using `self.findChild(QObject, 'name_of_the_combo')`

**Comment by ghtmtt** [comment]
> ```
> plotBaseProperties = {
>     'x': { 'default' : None, 'plot_types': ['all'] }
>     'y': { 'default' : None, 'plot_types': ['scatter', 'bar', 'box', 'pie', '2dhistogram', 'polar', 'ternary', 'contour'] },
>      'z': { 'default' : None, 'plot_types': ['ternary'] }
> }
> ```

actually this isn't that straightforward (but I have to think about that), because is you want to plot horizontal bars (e.g. for histogram or boxplot) the x become the y.

**Comment by mdouchin** [comment]
I added a commit with several improvements, among

**Comment by ghtmtt** [comment]
running some tests I noticed some problem with `box` and `histogram`. When you don't define the `Y` parameter (as you made correctly optional) I get this error for boxplot:

```
Traceback (most recent call last): File "/home/matteo/.local/share/QGIS/QGIS3/profiles/matteo/python/plugins/DataPlotly/processing/dataplotly_algorithms.py", line 183, in processAlgorithm raise QgsProcessingException(msg) _core.QgsProcessingException: The chosen plot type needs a Y field ! 

Execution failed after 0.01 seconds
```

and this for histogram:

```
Traceback (most recent call last): File "/home/matteo/.local/share/QGIS/QGIS3/profiles/matteo/python/plugins/DataPlotly/processing/dataplotly_algorithms.py", line 220, in processAlgorithm 'y_title': y_title UnboundLocalError: local variable 'y_title' referenced before assignment 

Execution failed after 0.01 seconds
```

I also noticed that when defining a path for the json output it is correctly handled, but when it is save as a temporary file nothing is generated anywhere. Obviously it make sense to not load the json in QGIS, but maybe we can add a title like `to see the file you have to choose a path` or something like that to give the user the idea.

One last thing. The optional Color parameter is great, but the user won't have any idea how to choose a different color. Typing a color (e.g. red, green, black) works nice. What about having a combobox with a small list of colors? I don't know if Processing can handle color widgets (I don't think so), so a short list could be an idea. 

Thanks for the huge work, really really appreciated!

**Comment by mdouchin** [comment]
List of colors changed in last  commit. I set opacity to 0.8 to avoid these standard colors to be too strong

**Comment by ghtmtt** [comment]
mm sorry but I'm not so in favor of setting a default opacity of 0.8. This mean changing also the UI (slider and number): even if not a problem I'd prefer to keep it to the default value.

**Comment by mdouchin** [comment]
* Opacity set back to None
* For X and Y: 
  - X is now optionnal
  - I added lists of plots mandatory for each dimension. I followed Dataplotly dock options for X and Y.
  - The error is set with processing "good practices", which means the alg raises an Exception. I also use the `feedback`

**Comment by mdouchin** [comment]
I added an output for the JSON file, but processing does not show it in the result viewer. I don't know why

**Comment by mdouchin** [comment]
198e7f5 adds a new input which allows to choose a HTML text color, which overrides the color chosen in the list. This will help advanced users to choose any color, and not force us to fill the color list with too many choices.

**Comment by ghtmtt** [comment]
@mdouchin this PR is awesome! I made some random tests with different plot type, layers etc.. and all seems perfectly working. Just one minor thing (but we can work on this in a **second moment**) is that if the algorithm is used within a model the plot output is not visible in the Result Viewer (I think it could be something related to the model itself).

So thanks again for taking care of this tremendous improvement! I'm going to merge it. 

I hope to release this week the updated release of the plugin (I've already changed the javascript file according to the 1.32 plotly release, see https://github.com/plotly/plotly.js/releases/tag/v1.32.0. They released also the 1.33 (the run like hell :) ), but I'll wait a while. Furthermore I'm waiting for an updated version of python plotly (currently I'm working with `2.0.13` but the `2.2.3` can be downloaded https://pypi.python.org/pypi/plotly). I'm not sure (but almost) that `2.2.3` does not contain `1.32` or `1.33` changed (like violin plots, the new polar plot...) so the current js in the repo is just for some *cosmetic* stuff (e.g. multiple selection on plot).

For this reason I'm quite conservative to update the python library and the related plotly plugin API (especially because the updated version should be loaded in the OSGEO installer). 

Thanks again for this work!

**Comment by mdouchin** [comment]
Thanks a lot for your kind review. I am happy to contribute to this great plugin !

About the JSON output not shown in the result viewer, I have no clue why, since I use the exact same addOutput method in the alg. Probably a QGIS "bug".

**Comment by ghtmtt** [comment]
maybe it is worth to forward this also in the qgis dev list. Might be a quick fix ;)

**Pull request merged**